tools/ui-components/src/helpers.tsx

```diff
@@ -0,0 +1,34 @@
+import React, { Context } from 'react';
+
+// I had to alter this to  < React.ComponentPropsWithRef<Inner>, number> and I don't know why
+export type ReplaceProps<Inner extends React.ElementType, P> = Omit<
+  React.ComponentPropsWithRef<Inner>,
+  number
+> &
+  P;
+
+export interface AsProp<As extends React.ElementType = React.ElementType> {
+  as?: As;
+}
+
+export interface BsPrefixOnlyProps {
+  bsPrefix?: string;
+}
+
+export interface BsPrefixProps<As extends React.ElementType = React.ElementType>
+  extends BsPrefixOnlyProps,
+    AsProp<As> {}
+
+export interface BsPrefixRefForwardingComponent<
+  TInitial extends React.ElementType,
+  P = unknown
+> {
+  <As extends React.ElementType = TInitial>(
+    props: React.PropsWithChildren<ReplaceProps<As, BsPrefixProps<As> & P>>,
+    context?: Context<string>
+  ): React.ReactElement | null;
+  propTypes?: React.ReactNode;
+  contextTypes?: Context<string>;
+  defaultProps?: Partial<P>;
+  displayName?: string;
+}
```

Comment:
I have kept the helpers because I don't know if they are needed in the future form, but I am planning to delete them and use what's only needed in their specific components

Reply:
Essentially, anything that is not used can be deleted. If we need it in the future, we can bring it back.